### ISSUE TITLE
npm install isn't including a sub dependency if it is also a main level devDependency with a different version

### ISSUE THREAD

**Issue by jamieotter** [npm/npm#10277]
This has all been verified with npm 3.3.10 and 3.3.12.  What we're seeing is that if you run `npm install` on the below package.json, `nject@2.0.1` doesn't get installed under `express-train` as expected.

```
{
  "dependencies": {
    "express-train": "~4.0.0"
  },
  "devDependencies": {
    "nject": "~1.3.2"
  }
}
```

This is the npm output, which is consistent with what we see in `node_modules` (note: no mention of `nject@2.0.1` under `express-train@4.0.1` or anywhere else).

```
├─┬ express-train@4.0.1
│ ├─┬ glob@4.0.6
│ │ ├── graceful-fs@3.0.8
│ │ ├── inherits@2.0.1
│ │ ├─┬ minimatch@1.0.0
│ │ │ ├── lru-cache@2.7.0
│ │ │ └── sigmund@1.0.1
│ │ └─┬ once@1.3.2
│ │   └── wrappy@1.0.1
│ ├─┬ handlebars@1.0.12
│ │ ├─┬ optimist@0.3.7
│ │ │ └── wordwrap@0.0.3
│ │ └─┬ uglify-js@2.3.6
│ │   ├── async@0.2.10
│ │   └─┬ source-map@0.1.43
│ │     └── amdefine@1.0.0
│ └── lodash@2.2.1
└─┬ nject@1.3.2
  └── lodash@2.3.0
```

`nject@2.0.1` is a dependent of `express-train@4.0.1`, but we also have a main level devDependency of `nject@1.3.2`.  The expectation is that `nject@2.0.1` would get installed under `express-train` and `nject@1.3.2` would end up installed on the root.

If we run `npm install` a second time things end up as expected with `nject@2.0.1` present under `express-train` this time around.  This is the output of `npm install` when run a second time:

```
├─┬ express-train@4.0.1
│ └─┬ nject@2.0.1
│   └── lodash@2.3.0
└─┬ nject@1.3.2
  └── lodash@2.3.0
```

Also, if we move `nject@1.3.2` from devDependencies to regular dependencies things will come out as expected.  Another solution is to `npm uninstall express-train` after the initial `npm install`, then `npm install` again, and things will also come out correctly.

**Comment by chrishiestand** [comment]
This may not be a realistic solution because I believe it would require changes to node itself, but in order to truly flatten dependencies, we could make dependency folders in the format: `node_modules/$dependency/$version/$dependency_$version_root`

So in the example above the project root would contain both:

    node_modules/nject/2.0.1/#2.0.1 module code
    node_modules/nject/1.3.2/#1.3.2 module code

When a module is imported, parsing of each package.json would be required to find the correct path, which may have startup performance penalties.

**Comment by lxe** [comment]
> Also, if we move nject@1.3.2 from devDependencies to regular dependencies things will come out as expected.

This workaround has worked for me as well :+1:

**Comment by schmod** [comment]
This looks like a dupe of #10727

**Comment by zkat** [comment]
Definitely looks that way, @schmod. I've also pinged about your PR, thanks! I'm gonna close this as a duplicate of https://github.com/npm/npm/issues/10727

**Comment by zkat** [comment]
This is quite probably fixed as of npm@4.0.5 thanks to #12811. Cheers!
